src/datasets/demo_replay.py

```diff
@@ -70,8 +84,63 @@ def create_demo_replay_data(env: BaseEnv, train_tasks: List[Task]) -> Dataset:
         except EnvironmentFailure:
             # We ignore replay data which leads to an environment failure.
             continue
+
+        if nonoptimal_only and _replay_is_optimal(replay_traj, traj, state_idx,
+                                                  oracle_approach, env):
+            continue
+
         if CFG.option_learner != "no_learning":
             for act in replay_traj.actions:
                 act.unset_option()
         replay_dataset.append(replay_traj)
+
+    assert len(replay_dataset) == CFG.offline_data_num_replays
+
     return demo_dataset + replay_dataset
+
+
+def _replay_is_optimal(replay_traj: LowLevelTrajectory,
+                       demo_traj: LowLevelTrajectory, state_idx: int,
+                       oracle_approach: BaseApproach, env: BaseEnv) -> bool:
+    """Plan from the end of the replay to the goal and check whether the result
+    is as good as the demo.
+
+    The state_idx is the index into demo_traj.states where the replay
+    started.
+    """
+    assert demo_traj.states[state_idx].allclose(replay_traj.states[0])
+    # Plan starting at the end of the replay trajectory to the demo goal.
+    task = Task(replay_traj.states[-1], demo_traj.goal)
+    try:
+        policy = oracle_approach.solve(
+            task, timeout=CFG.offline_data_planning_timeout)
```

Comment:
won't we have to change this to like 3 in order for this whole thing to actually run? maybe let's just do that now?

Reply:
ah okay you already did